### PR TITLE
[cdc] Update flink.cdc.version to 3.5.0 and add FlinkCDC to Paimon converter.

### PR DESCRIPTION
**Pull request by lvyanquan**
### Purpose

Linked issue: close #xxx

### CODE REVIEW

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/util/FlinkCDCToPaimonTypeConverter.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                                builder.column(
+                                        column.getName(),
+                                        convertFlinkCDCDataTypeToPaimonDataType(column.getType()),
+                                        column.getComment(),
+                                        convertFlinkCDCDefaultValueToValidValue(
+                                                column.getDefaultValueExpression(),
+                                                column.getType())));
+        builder.primaryKey(schema.primaryKeys())
+                .partitionKeys(schema.partitionKeys())
+                .comment(schema.comment())
```
Comment (yunfengzhou-hub) [inline paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/util/FlinkCDCToPaimonTypeConverter.java]: Here need to check whether comment is null, otherwise might throw exceptions

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/util/PaimonToFlinkCDCDataConverter.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        Object[] objects = new Object[row.getFieldCount()];
+        for (int i = 0; i < row.getFieldCount(); i++) {
+            objects[i] = fieldGetters.get(i).getFieldOrNull(row);
+        }
+        BinaryRecordData binaryRecordData = recordDataGenerator.generate(objects);
+        switch (row.getRowKind()) {
+            case INSERT:
+            case UPDATE_AFTER:
+                {
+                    return DataChangeEvent.insertEvent(tableId, binaryRecordData, new HashMap<>());
```
Comment (yunfengzhou-hub) [inline paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/util/PaimonToFlinkCDCDataConverter.java:56]: Should it be `DataChangeEvent.updateEvent`? Same for the `UPDATE_BEFORE` below.

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/util/PaimonToFlinkCDCDataConverter.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+                }
+            case DELETE:
+            case UPDATE_BEFORE:
+                {
+                    return DataChangeEvent.deleteEvent(tableId, binaryRecordData, new HashMap<>());
+                }
+            default:
+                throw new IllegalArgumentException("don't support type of " + row.getRowKind());
```
Comment (yunfengzhou-hub) [inline paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/util/PaimonToFlinkCDCDataConverter.java:64]: nit: Capitalize the first letter of don't. Same as `FlinkCDCToPaimonDataConverter`.

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/util/PaimonToFlinkCDCDataConverter.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+                fieldGetter = row -> row.getFloat(fieldPos);
+                break;
+            case DOUBLE:
+                fieldGetter = row -> row.getDouble(fieldPos);
+                break;
+            case INTEGER:
+                fieldGetter = row -> row.getInt(fieldPos);
+                break;
+            case DATE:
+                fieldGetter = row -> row.getInt(fieldPos);
```
Comment (yunfengzhou-hub) [inline paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/util/PaimonToFlinkCDCDataConverter.java:122]: Should the result type be something like DateTime or Instant? Same for below

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/util/PaimonToFlinkCDCDataConverter.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+                                                        fieldPos,
+                                                        DataTypeChecks.getPrecision(fieldType))
+                                                .toInstant());
+                break;
+            default:
+                throw new IllegalArgumentException(
+                        "don't support type of " + fieldType.getTypeRoot());
+        }
+        if (!fieldType.isNullable()) {
+            return fieldGetter;
```
Comment (yunfengzhou-hub) [inline paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/util/PaimonToFlinkCDCDataConverter.java:150]: Let's add a test to verify the behavior when values are null but types require not null.

File: paimon-flink/paimon-flink-cdc/src/test/java/org/apache/paimon/flink/pipeline/cdc/util/DataConvertTest.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+                    TimestampData.fromMillis(400, 4),
+                    LocalZonedTimestampData.fromEpochMillis(300, 3),
+                    LocalZonedTimestampData.fromEpochMillis(400, 4),
+                    LocalZonedTimestampData.fromEpochMillis(500, 5),
+                    LocalZonedTimestampData.fromEpochMillis(600, 6),
+                };
+        org.apache.flink.cdc.common.event.DataChangeEvent dataChangeEvent =
+                DataChangeEvent.insertEvent(tableId, recordDataGenerator.generate(testData));
+
+        Assertions.assertEquals(
```
Comment (yunfengzhou-hub) [inline paimon-flink/paimon-flink-cdc/src/test/java/org/apache/paimon/flink/pipeline/cdc/util/DataConvertTest.java:145]: Let's also verify the converted data in the middle phase, after the original CDC data is converted to paimon data and before it's converted back.

File: paimon-flink/paimon-flink-cdc/src/test/java/org/apache/paimon/flink/pipeline/cdc/util/TypeConverterTest.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+                                "timestamp_ltz_with_precision_9",
+                                org.apache.flink.cdc.common.types.DataTypes.TIMESTAMP_LTZ(9))
+                        .primaryKey("pk_string")
+                        .partitionKey("boolean")
+                        .build();
+
+        Assertions.assertEquals(
+                fullTypesSchema,
+                PaimonToFlinkCDCTypeConverter.convertPaimonSchemaToFlinkCDCSchema(
+                        FlinkCDCToPaimonTypeConverter.convertFlinkCDCSchemaToPaimonSchema(
```
Comment (yunfengzhou-hub) [inline paimon-flink/paimon-flink-cdc/src/test/java/org/apache/paimon/flink/pipeline/cdc/util/TypeConverterTest.java:101]: Same as above.

File: paimon-flink/paimon-flink-cdc/src/test/resources/log4j2-test.properties [hunk 8]
```diff
@@ -18,7 +18,7 @@
 
 # Set root logger level to OFF to not flood build logs
 # set manually to INFO for debugging purposes
-rootLogger.level = OFF
+rootLogger.level = INFO
```
Comment (yunfengzhou-hub) [inline paimon-flink/paimon-flink-cdc/src/test/resources/log4j2-test.properties:21]: Maybe need to revert this change.

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/schema/PaimonMetadataApplier.java [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+                                                            .getLogicalType()),
+                                            column.getComment()));
+            List<String> partitionKeys = new ArrayList<>();
+            List<String> primaryKeys = schema.primaryKeys();
+            if (partitionMaps.containsKey(event.tableId())) {
+                partitionKeys.addAll(partitionMaps.get(event.tableId()));
+            } else if (schema.partitionKeys() != null && !schema.partitionKeys().isEmpty()) {
+                partitionKeys.addAll(schema.partitionKeys());
+            }
+            builder.primaryKey(primaryKeys)
```
Comment (yunfengzhou-hub) [inline paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/schema/PaimonMetadataApplier.java:184]: I noticed that in Flink CDC, this implementation originally looks like this
```
            for (String partitionColumn : partitionKeys) {
                if (!primaryKeys.contains(partitionColumn)) {
                    primaryKeys.add(partitionColumn);
                }
            }
            builder.partitionKeys(partitionKeys)
                    .primaryKey(primaryKeys)
```

Why do we change the implementation here?

File: paimon-flink/paimon-flink-cdc/src/test/java/org/apache/paimon/flink/pipeline/cdc/schema/PaimonMetadataApplierTest.java [hunk 10]
```diff
@@ -0,0 +1,609 @@
...
+                                new DataField(3, "col4", DataTypes.STRING(), "comment of col4")));
+        Assertions.assertThat(table.rowType()).isEqualTo(tableSchema);
+        Assertions.assertThat(table.primaryKeys()).isEmpty();
+        Assertions.assertThat(table.partitionKeys()).isEmpty();
+        Assertions.assertThat(table.options()).containsEntry("bucket", "-1");
+        Assertions.assertThat(table.comment()).contains("comment of table_with_comment");
+    }
+
+    @Test
+    public void testMysqlDefaultTimestampValueConversionInAddColumn()
```
Comment (yunfengzhou-hub) [inline paimon-flink/paimon-flink-cdc/src/test/java/org/apache/paimon/flink/pipeline/cdc/schema/PaimonMetadataApplierTest.java:567]: Is "Mysql" a typo? Or maybe better to add a comment describing why this paimon test is related to mysql.

File: paimon-flink/paimon-flink-cdc/src/test/java/org/apache/paimon/flink/action/cdc/mongodb/MongoDBSyncDatabaseActionITCase.java [hunk 11]
```diff
@@ -298,7 +298,7 @@ public void testNewlyAddedTablesOptionsChange() throws Exception {
         JobClient jobClient = runActionWithDefaultEnv(action1);
 
         waitingTables("t3");
-        jobClient.cancel();
+        jobClient.cancel().get();
```
Comment (yunfengzhou-hub) [inline paimon-flink/paimon-flink-cdc/src/test/java/org/apache/paimon/flink/action/cdc/mongodb/MongoDBSyncDatabaseActionITCase.java:301]: Better add a timeout to avoid infinite blocking. Same for other test cases.

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/schema/SchemaChangeProvider.java [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+    /**
+     * Creates a SchemaChange object for setting an option.
+     *
+     * @param key The key of the option to be set.
+     * @param value The value of the option to be set.
+     * @return A SchemaChange object representing the setting of an option.
+     */
+    public static SchemaChange setOption(String key, String value) {
```
Comment (yunfengzhou-hub) [inline paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/schema/SchemaChangeProvider.java:149]: unused method.

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/postgres/PostgresActionUtils.java [hunk 13]
```diff
@@ -186,7 +186,7 @@ public static JdbcIncrementalSource<CdcSourceRecord> buildPostgresSource(
         customConverterConfigs.put(JsonConverterConfig.DECIMAL_FORMAT_CONFIG, "numeric");
         CdcDebeziumDeserializationSchema schema =
                 new CdcDebeziumDeserializationSchema(true, customConverterConfigs);
-        return sourceBuilder.deserializer(schema).includeSchemaChanges(true).build();
+        return sourceBuilder.deserializer(schema).build();
```
Comment (yunfengzhou-hub) [inline paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/postgres/PostgresActionUtils.java:189]: Shall we describe why we need to make changes like this in non-Paimon-Flink-CDC classes? I suppose these changes are along with the Flink CDC version upgrade from 3.1 to 3.5, but not quite sure why we need to make such changes. We can add some details to the description section of this PR.

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/util/PaimonToFlinkCDCDataConverter.java (hunk 2 above)
Comment (yuxiqian) [reply paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/util/PaimonToFlinkCDCDataConverter.java:56]: Flink Row encodes an update event as two separated rows (`-U` and `+U`), while CDC `UpdateEvent` carries both as one event. Seems it's not possible to build a complete `DataChangeEvent.updateEvent` with given `UPDATE_BEFORE` or `UPDATE_AFTER` row here.

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/util/PaimonToFlinkCDCDataConverter.java (hunk 5 above)
Comment (yuxiqian) [reply paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/util/PaimonToFlinkCDCDataConverter.java:150]: Some quick experiments show that a "default" value will be filled when trying to write a `null` value into NOT NULL fields, like `false` for `BOOLEAN` and `1970-01-01` for `DATE`.

Shall we a) keep this behavior and add test to cover this behavior, or b) validate nullability violations and throw an exception if occurs?

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/schema/PaimonMetadataApplier.java (hunk 9 above)
Comment (yuxiqian) [reply paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/pipeline/cdc/schema/PaimonMetadataApplier.java:184]: Added it back for consistency.

File: paimon-flink/paimon-flink-cdc/src/test/java/org/apache/paimon/flink/pipeline/cdc/schema/PaimonMetadataApplierTest.java (hunk 10 above)
Comment (yuxiqian) [reply paimon-flink/paimon-flink-cdc/src/test/java/org/apache/paimon/flink/pipeline/cdc/schema/PaimonMetadataApplierTest.java:567]: I think the original test case is meant to cover some special default values in MySQL timestamp fields (like `CURRENT_TIMESTAMP`). Renamed it to avoid confusion.

File: paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/postgres/PostgresActionUtils.java (hunk 13 above)
Comment (yuxiqian) [reply paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/action/cdc/postgres/PostgresActionUtils.java:189]: IIUC due to technical limitations, PostgreSQL CDC does not support capturing DDL events or emitting any schema change events. In earlier versions this method (`includeSchemaChanges`) is presented but has no effect at all, and got removed in https://github.com/apache/flink-cdc/pull/3464.
